pkg/foundation/grpcutil/websocket.go

```diff
@@ -0,0 +1,132 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcutil
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"net/http"
+
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/gorilla/websocket"
+)
+
+type inMemoryResponseWriter struct {
+	io.Writer
+	header http.Header
+	closed chan bool
+}
+
+func newInMemoryResponseWriter(writer io.Writer) *inMemoryResponseWriter {
+	return &inMemoryResponseWriter{
+		Writer: writer,
+		header: http.Header{},
+		closed: make(chan bool, 1),
+	}
+}
+
+func (w *inMemoryResponseWriter) Write(b []byte) (int, error) {
+	return w.Writer.Write(b)
+}
+func (w *inMemoryResponseWriter) Header() http.Header {
+	return w.header
+}
+func (w *inMemoryResponseWriter) WriteHeader(int) {
+	// we don't have a use for the code
+}
+func (w *inMemoryResponseWriter) CloseNotify() <-chan bool {
+	return w.closed
+}
+func (w *inMemoryResponseWriter) Flush() {}
+
+// wsProxy is a proxy around a http.Handler which
+// redirects the response data from the http.Handler
+// to a WebSocket connection.
+type wsProxy struct {
+	handler  http.Handler
+	logger   log.CtxLogger
+	upgrader websocket.Upgrader
+}
+
+func newWebSocketProxy(handler http.Handler, logger log.CtxLogger) *wsProxy {
+	return &wsProxy{
+		handler: handler,
+		logger:  logger.WithComponent("grpcutil.websocket"),
+		upgrader: websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+		},
```

Comment:
We should allow any origin, otherwise browsers won't be able to connect to the websocket (e.g. websocket king complains it can't connect).
```suggestion
		upgrader: websocket.Upgrader{
			ReadBufferSize:  1024,
			WriteBufferSize: 1024,
			CheckOrigin:     func(r *http.Request) bool { return true },
		},
```

Reply:
Yeah, I noticed that in the original implementation. I've intentionally removed because checking the origin should generally protect against CSRF. CLI tools and our UI continues to work with the checked origins, so it's only web sites like WebSocketKing which have problems and I think it's an acceptable trade off for the security.